File: examples/word-replace.ts
/**
 * 例子：打开 docx，把每个 `{{client}}` 占位文本替换为指定值，写回新文件。
 *
 * 跑法：
 *   bun run examples/word-replace.ts <input.docx> <output.docx> [<client-name>]
 */

import { Text, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [inputPath, outputPath, client = "Acme Corp"] = process.argv.slice(2);
  if (inputPath === undefined || outputPath === undefined) {
    process.stderr.write("usage: word-replace <input.docx> <output.docx> [<client-name>]\n");
    process.exit(2);
  }

  const doc = await WordprocessingDocument.openAsync(inputPath);
  const body = doc.mainDocumentPart?.document;
  if (body === undefined) {
    process.stderr.write("error: input is not a valid docx (no main document part)\n");
    process.exit(1);
  }

  let replaced = 0;
  for (const t of body.descendants(Text)) {
    if (t.text?.includes("{{client}}")) {
      t.text = t.text.replaceAll("{{client}}", client);
      replaced += 1;
    }
  }

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Replaced ${replaced} occurrence(s); wrote ${outputPath}\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordReplace.cs
// Replica of examples/word-replace.ts
// Creates a source docx with "{{client}}" placeholder, opens it, replaces it,
// and saves to outputPath — equivalent to running word-replace.ts on a generated input.
//
// Since word-replace.ts is a transform (read input, write output), we generate
// the source document inline and then apply the replacement, producing the same
// final output as: bun run word-replace.ts <input> <output> "Acme Corp"

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordReplace
{
    private const string Client = "Acme Corp";

    public static void Run(string outputPath)
    {
        // Step 1: build the "template" document in memory via a temp file
        var tempPath = Path.GetTempFileName() + ".docx";
        try
        {
            BuildTemplate(tempPath);
            ApplyReplace(tempPath, outputPath);
        }
        finally
        {
            if (File.Exists(tempPath)) File.Delete(tempPath);
        }
    }

    private static void BuildTemplate(string path)
    {
        using var doc = WordprocessingDocument.Create(path,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);
        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        body.Append(new Paragraph(new Run(new Text("Dear {{client}},"))));
        body.Append(new Paragraph(new Run(new Text("Thank you for your business, {{client}}."))));
        body.Append(new Paragraph(new Run(new Text("Sincerely,"))));
        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }

    private static void ApplyReplace(string inputPath, string outputPath)
    {
        // Copy to output first (we open in-place)
        File.Copy(inputPath, outputPath, overwrite: true);

        using var doc = WordprocessingDocument.Open(outputPath, true);
        var body = doc.MainDocumentPart?.Document?.Body;
        if (body == null) return;

        foreach (var t in body.Descendants<Text>())
        {
            if (t.Text.Contains("{{client}}"))
                t.Text = t.Text.Replace("{{client}}", Client);
        }

        doc.MainDocumentPart!.Document.Save();
    }
}
